pkg/agent/core/ngt/service/ngt.go

```diff
@@ -569,89 +542,46 @@ func (n *ngt) CreateIndex(ctx context.Context, poolSize uint32) (err error) {
 	if n.IsIndexing() || n.IsSaving() {
 		return nil
 	}
-	ic := atomic.LoadUint64(&n.ic)
+	ic := n.vq.IVQLen() + n.vq.DVQLen()
 	if ic == 0 {
 		return errors.ErrUncommittedIndexNotFound
 	}
 	n.indexing.Store(true)
-	atomic.StoreUint64(&n.ic, 0)
-	t := time.Now().UnixNano()
 	defer n.indexing.Store(false)
 	defer n.gc()
 
 	log.Infof("create index operation started, uncommitted indexes = %d", ic)
-	delList := make([]string, 0, ic)
-	n.dvc.Range(func(uuid string, dvc vcache) bool {
-		if dvc.date > t {
-			return true
-		}
-		if ivc, ok := n.ivc.Load(uuid); ok && ivc.date < t && ivc.date < dvc.date {
-			n.ivc.Delete(uuid)
+	log.Info("create index delete phase started")
+	n.vq.RangePopDelete(ctx, func(uuid string) bool {
+		var ierr error
+		oid, ok := n.kvs.Delete(uuid)
+		if ok {
+			ierr = n.core.Remove(uint(oid))
+			if ierr != nil {
+				log.Error(ierr)
+				err = errors.Wrap(err, ierr.Error())
+			}
+		} else {
+			ierr = errors.ErrObjectIDNotFound(uuid)
+			log.Error(ierr)
+			err = errors.Wrap(err, ierr.Error())
 		}
-		delList = append(delList, uuid)
 		return true
 	})
-	log.Info("create index delete kvs phase started")
-	log.Debugf("deleting kvs: %#v", delList)
-	doids := make([]uint, 0, ic)
-	for _, duuid := range delList {
-		n.dvc.Delete(duuid)
-		id, ok := n.kvs.Delete(duuid)
-		if !ok {
-			log.Error(errors.ErrObjectIDNotFound(duuid).Error())
-			err = errors.Wrap(err, errors.ErrObjectIDNotFound(duuid).Error())
+	log.Info("create index delete phase finished")
+	n.gc()
+	log.Info("create index insert phase started")
+	n.vq.RangePopInsert(ctx, func(uuid string, vector []float32) bool {
```

Comment:
I am not sure if it is correct:
The current logic is process all the delete request first, and insert request later

For example we have 2 request, reqA and reqB
reqA: insert vecA request, time t+0
reqB: delete vecA request, time t+10

If they are processing the same vector, the final result will be the vecA is inserted, but maybe it is not correct.

Am I correct?

Reply:
that's why I remove the Insert Queue here.
https://github.com/vdaas/vald/pull/1028/files#diff-d8459a99ea4f8ec96e01405d5f55171070363986154ad61e269c7911dbfd91a9R304-R327

Reply:
I see... I think it is good.